File: ChatGPT/Fáceis/dfs/dfs.java
// Input  Using the previous graph, how to implement the Depth-First Search algorith?
import java.util.ArrayList;
import java.util.LinkedList;

public class Graph {
    private ArrayList<LinkedList<Integer>> adjList;
    private int numVertices;

    public Graph(int numVertices) {
        this.numVertices = numVertices;
        adjList = new ArrayList<>(numVertices);
        for (int i = 0; i < numVertices; i++) {
            adjList.add(new LinkedList<>());
        }
    }

    // Other methods (addEdge, removeEdge, etc.)

    public void dfs(int startVertex) {
        boolean[] visited = new boolean[numVertices];
        dfsRecursive(startVertex, visited);
    }

    private void dfsRecursive(int vertex, boolean[] visited) {
        visited[vertex] = true;
        System.out.print(vertex + " "); // Process the vertex

        for (int neighbor : adjList.get(vertex)) {
            if (!visited[neighbor]) { // Check if not visited
                dfsRecursive(neighbor, visited);
            }
        }
    }
}

File: ChatGPT/Simples/dfs/dfs.cpp
// Input: Using the previous graph, how to implement the Depth-First Search algorith?

#include <iostream>
#include <vector>
#include <list>

class Graph {
private:
    std::vector<std::list<int>> adjList;
    int numVertices;

    void DFSUtil(int v, std::vector<bool>& visited) {
        // Mark the current node as visited
        visited[v] = true;
        std::cout << v << " ";

        // Recur for all adjacent vertices
        for (const int& vertex : adjList[v]) {
            if (!visited[vertex]) {
                DFSUtil(vertex, visited);
            }
        }
    }

public:
    Graph(int vertices) : numVertices(vertices) {
        adjList.resize(vertices);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u); // For undirected graph
    }

    void DFS(int start) {
        std::vector<bool> visited(numVertices, false);
        DFSUtil(start, visited);
    }
};

int main() {
    Graph g(4);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 3);

    std::cout << "DFS starting from vertex 0:" << std::endl;
    g.DFS(0);

    return 0;
}

/*
Score: 1
*/